### DISCUSSION TITLE
Postgresql supports fetching tuples

### DISCUSSION THREAD

**Discussion by CaselIT** [sqlalchemy/sqlalchemy#9951] [Blue Sky]
This may go as one of the strange PG only features, but a query like this
```sql
SELECT (4,2)
```
is perfectly legal in postgresql, but it's not supported by sqlalchemy, unless run in textual form.
It trips up the compiler and ignoring the compiler error leads to NotImplementedError in tuple result_processor.

Also it seems that driver support if quite spotty, with only asyncpg behaving correctly:
- psycopg2 returns the string `'(4,2)'`
- psycopg returns the tuple `('4','2')` (strings instead of ints)
- pg8000 returns the string `'(4,2)'`
- asyncpg returns the correct tuple `(4,2)`

**Comment by tlocke** [comment]
Looking into this I found that PostgreSQL sends this as a `record` type. The PostgreSQL client protocol can use either text or binary data type representations. pg8000 can't properly parse a `record` because it uses the text representation and so doesn't know the types of the elements of the record. On the other hand asyncpg uses the binary representation of types, and this encoding does contain the types of the elements and so it can be parsed properly.

**Comment by CaselIT** [comment]
I guess `eval` could work in pg8000 😶
Jokes aside, I assumed something similar. There isn't a particular need for this feature, as evidenced by the missing support of the drivers in general, I just opened this since I noticed an error in the compiler for tuple that not completely correct for pg.

**Comment by tlocke** [comment]
Your comment made me think that pg8000 could do a bit better in this case, and so now, rather than return the entire raw string, [pg8000 returns](https://github.com/tlocke/pg8000/commit/2dd506f99edd7ed2c719e6ac558e9a41ed3e0ae9) a tuple of strings `('4','2')`.
